Apache Shiro based authentication

TimMaltGermany opened the pull request:
Implementation of user authentication via LDAP that utilizes the Apache Shiro framework.
Currently, two authentication mechanisms are supported:
1) Basic LDAP authentication with username / password without group membership checking
2) LDAP authentication with username / password and additional group membership checks based on the 'member' attribute of LDAP groups.

It should be relatively straight forward to add further authentication realms, but we chose to restrict the PR to LDAP only for now.

This implementation does not use session tokens and re-authenticates users with every request. Shiro does have a caching mechanism which is activated in the shiro.ini file.

Currently all but two routes require authentication. We chose not to require authentication for "otherRoutes" and "healthzRoutes" in WebApi.

Thanks,
the KNIME team. 

Last check-in comment:
- first authentication plug-in is LDAP, with and without group checking
 - addressed review comments from first pull-request
 - simplified configuration
 - added tests and documentation

zeitos commented on job-server/src/spark.jobserver/auth/LdapGroupRealm.scala:
```diff
@@ -0,0 +1,144 @@
+import org.apache.shiro.authz._
+import javax.naming.ldap.LdapContext
+import javax.naming.directory._
+import javax.naming.NamingEnumeration
```
this import is unused.

zeitos commented on job-server/src/spark.jobserver/auth/LdapGroupRealm.scala:
```diff
@@ -0,0 +1,144 @@
+  lazy val allowedGroups: Option[Array[String]] = getContextFactory() match {
+    case jni: JndiLdapContextFactory =>
+      val groups = getEnvironmentParam(jni, "ldap.allowedGroups").
+        split("\"").filter(group => group.replaceAll("[, ]", "").length() > 0)
```
nit: since " is just one char you can do '"' (... but I'm not sure if it will be more readable)

TimMaltGermany replied:
I personally prefer the more explicit slash.

zeitos commented on job-server/src/spark.jobserver/auth/LdapGroupRealm.scala:
```diff
@@ -0,0 +1,144 @@
+      groupSearchAtts, searchCtls).asScala
+
+    groupAnswer.map(sr2 => {
+      logger.debug("Checking members of group [%s]" format (sr2.getName()))
```
this line will create a new string for every groupAnswer even though debug level is not enabled.
Please check if level is enabled or do
```
logger.debug("Checking members of group [%s]" , sr2.getName())
```

zeitos commented on job-server/src/spark.jobserver/auth/LdapGroupRealm.scala:
```diff
@@ -0,0 +1,144 @@
+  def isInAllowedGroupOrNoCheckOnGroups(roles: Set[String]): Boolean = {
+    allowedGroups match {
+      case Some(groups) =>
+        roles.foldLeft(false)((isInGroup, r) => isInGroup || groups.contains(r))
```
you can do 
```
roles.exists(r => groups.contains(r))
```
that will stop looking once one is true

zeitos commented on job-server/src/spark.jobserver/auth/SJSAuthenticator.scala:
```diff
@@ -0,0 +1,101 @@
+      //  realm implementation may for example throw an exception
+      //  if user is not a member of a valid group
+      currentUser.isPermitted("*")
+      logger.trace("ACCESS GRANTED, user [%s]" format fullName)
```
same I did before with logger.debug

zeitos commented:
@TimMaltGermany Great work! I only added some minor cosmetic comments.

TimMaltGermany commented:
@zeitos Thanks for the hints. I changed the code as suggested and added a few more test cases.

TimMaltGermany commented:
So, does this mean that we are ready to merge or are we waiting for someone else to comment? 
Thanks,
Tim

zeitos commented:
@velvia ?

velvia commented:
@TimMaltGermany @zeitos will have a look, probably tomorrow.

noorul commented on job-server/test/spark.jobserver/auth/SJSAuthenticatorSpec.scala:
```diff
@@ -0,0 +1,138 @@
+  def actorRefFactory = system
+
+  //edit this with your real LDAP server information, just remember not to 
+  // check it in....
```
Is manual intervention required for running tests?

TimMaltGermany replied:
No, i just left this in in case someone wants to test this against a real LDAP server
The actual test uses the simple authorization procedure provided by the Shiro example.

velvia commented on job-server/config/shiro.ini:
```diff
@@ -0,0 +1,16 @@
+
+# use this for basic ldap authorization, without group checking
+activeDirectoryRealm = org.apache.shiro.realm.ldap.JndiLdapRealm
```
I'm wondering if this should be named `shiro.ini.template`, and adding to the deploy section of the README, because people could deploy with this and not realize they have to change the settings.

TimMaltGermany replied:
@velvia  sure, makes sense. How about this for the README.md then: 
1.1 Copy `config/local.sh.template` to `<environment>.sh` .....
1.2 Copy `config/shiro.ini.template` to `shiro.ini` and edit as appropriate. NOTE: only required when `authentication = on`

velvia commented on job-server/src/spark.jobserver/WebApi.scala:
```diff
@@ -51,6 +56,18 @@ class WebApi(system: ActorSystem,
+  lazy val authenticator: AuthMagnet[AuthInfo] = {
+    if (config.getBoolean("shiro.authentication")) {
+      logger.info("Using authentication.")
+      val sManager = new IniSecurityManagerFactory(config.getString("shiro.config.path")).getInstance()
```
Yes, this is much better  :)

velvia commented on job-server/src/spark.jobserver/auth/LdapGroupRealm.scala:
```diff
@@ -0,0 +1,142 @@
+    val groupAnswer = ldapContext.search(searchBase, LdapGroupRealm.groupMemberFilter,
+      groupSearchAtts, searchCtls).asScala
+
+    groupAnswer.map(sr2 => {
```
Minor nit:  Stylistically the idiom for Scala is usally like this:

    groupAnswer.map { sr2 =>
      logger.debug(....)
    }.toMap

velvia commented on job-server/src/spark.jobserver/auth/LdapGroupRealm.scala:
```diff
@@ -0,0 +1,142 @@
+
+    val members = retrieveGroups(ldapContext)
+
+    answer.map(userSearchResult => {
```
Same thing here, `.map { userSearchResult =>`

velvia commented on job-server/src/spark.jobserver/auth/SJSAuthenticator.scala:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    def authenticator(userPass: Option[UserPass]): Future[Option[AuthInfo]] = Future {
+      Await.result(validate(userPass), Duration.Inf)
```
Should the duration be configurable?

TimMaltGermany replied:
sure, I will add the config option shiro.authentication-timeout = 10 s

velvia commented:
@TimMaltGermany overall :+1: , just some minor style comments, plus:

- Have you guys tried deploying this?  Do the deploy scripts work?  (Not necessary to roll this into this PR, could be a separate one, as long as we track it)
- What I said about `shiro.ini.template`

Thanks!  Looking forward to this, and thinking of doing a bigger version bump  :)

TimMaltGermany commented:
@velvia Thanks for your comments. I added two template files (ldap and basic username/password), modified the deploy script, changed the readme.md and tested the deploy - works for me.

TimMaltGermany commented:
@velvia give me a few more minutes before you merge so that I can add one more test class

TimMaltGermany commented:
@velvia Done

velvia commented:
@TimMaltGermany awesome, thanks!  Going to merge this.
